src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -15,93 +15,137 @@
-      aggregateValues.put(currentRange, 0);
+      int currentFrequency = 0;
+      PriorityQueue<UserComment> highMagnitudeComments = new PriorityQueue<>();
       // loop through sorted scores within currentRange from updated score pointer and update its
-      // corresponding appearance frequency in aggregatedValues
+      // corresponding appearance frequency
       int scoreIdx;
-      for (scoreIdx = updatingScoreIdx; scoreIdx < sentimentScores.size(); scoreIdx++) {
-        BigDecimal scorePoint = BigDecimal.valueOf(sentimentScores.get(scoreIdx));
-        if ((scorePoint.compareTo(nextPoint) < 0) || nextPoint.compareTo(UPPER_END) == 0) {
-          aggregateValues.put(currentRange, aggregateValues.get(currentRange) + 1);
+      for (scoreIdx = updatingScoreIdx; scoreIdx < userCommentList.size(); scoreIdx++) {
+        BigDecimal scorePoint = BigDecimal.valueOf(userCommentList.get(scoreIdx).getScore());
+        if ((scorePoint.compareTo(nextPoint) < 0) || nextPoint.compareTo(upperEnd) == 0) {
+          currentFrequency += 1;
+          // TODO: add topNumComments to the priority queue highMagnitudeList
         } else {
           break;
         }
       }
       // update the score pointer
       updatingScoreIdx = scoreIdx;
+      sentimentBucketList.add(
+          new SentimentBucket(
+              new ArrayList<>(highMagnitudeComments), currentFrequency, currentRange));
     }
+    return sentimentBucketList;
+  }
+
+  /**
+   * Set the average value of given sentiment values.
+   *
+   * @param userCommentList a list of userComment to calculate the average score for
+   * @return the average score of userCommentList
+   */
+  private double getAverageScore(List<UserComment> userCommentList) {
+    return userCommentList.parallelStream()
+        .mapToDouble(UserComment::getScore)
+        .average()
+        .orElseThrow(
+            () ->
+                new RuntimeException("Unable to calculate average score due to empty input list."));
   }
 
   /**
-   * Set the average score of given sentiment scores. Throws an runtime exception if the average
-   * score is not valid or none of the sentiment scores is valid.
+   * Set the average value of given sentiment magnitude.
    *
-   * @param sentimentScores a list of score values from -1.0 to 1.0
+   * @param userCommentList a list of userComment to calculate the average magnitude for
+   * @return the average magnitude of userCommentList
    */
-  private void setAverageScore(List<Double> sentimentScores) throws RuntimeException {
-    averageScore =
-        sentimentScores.stream()
-            .mapToDouble(i -> i)
-            .average()
-            .orElseThrow(
-                () ->
-                    new RuntimeException(
-                        "Unable to calculate sentiment average due to empty input list."));
+  private double getAverageMagnitude(List<UserComment> userCommentList) {
+    return userCommentList.parallelStream()
```

Comment:
Generally doing things in parallel requires more overhead than doing things serially since there are a lot more things that need to be done. E.g. it has to split the work required into different buckets, create new threads, mrege the output of the different threads, cleanup threads, etc.

For simple operations like `mapToDouble` it will likely be faster to not execute the stream in parallel.

So generally, it's best to invoke opeartions in parallel if there are large periods of time where you expect to be waiting (i.e. when you're sending an HTTP request to an API and are waiting for a response) or when the operation you're invoking is computationally expensive and can easily be made parallelizable.